src/lib/base/container.ts

```diff
@@ -0,0 +1,83 @@
+import BaseObject from './base-object';
+import { AlreadyContains, CannotContain, CannotRelease } from './exceptions';
+
+/**
+ * A class representing a container, which can hold things
+ */
+export default class Container extends BaseObject {
+  protected _contents: Set<BaseObject> = new Set([]);
+  public maxContentCount: number = Infinity;
+  public maxContentWeight: number = Infinity;
+
+  get contents(): Array<BaseObject> {
+    return [...this._contents];
+  }
+
+  get contentWeight(): number {
+    return this.contents.reduce((total, item) => total + item.weight, 0.0);
+  }
+
+  get weight(): number {
+    return this._weight + this.contentWeight;
+  }
+
+  set weight(weight) {
+    super.weight = weight;
+  }
+
+  /**
+   * Check whether this contains another object
+   */
+  contains(obj: BaseObject): boolean {
+    return this._contents.has(obj);
+  }
+
+  /**
+   * Check whether obj can be put inside this container
+   */
+  canContain(obj: BaseObject): boolean {
+    // FIXME: It seems like throw/catch is not a recommended pattern in JS.
```

Comment:
It is a recommended pattern in JS, especially with async/await.
It seems like there is a ton of debate on exception handling in TS. IMO, it is something that should be addressed at the language level, rather than via a bunch of wrapping code to make errors typesafe (as in the example linked below). Personally, I lean more towards something like [this](https://github.com/microsoft/TypeScript/issues/13219).

Reply:
Well that's good to hear. I think I had run across that same issue previously, but with it still being open and such a lively debate, it was seeming to me like it wasn't recommended.

That said, the linked issue is hurting my head right now, and I'm just not going to bother implementing all of that. Seems like way too much work at the moment. :smile: